feat: voice recording

TrapDrap opened the pull request:
## Summary

This PR aims to implement voice recording to nextcord.
This is mostly fresh code, with some decoding & decrypting parts taken from the previous dead PR on voice recording.
Note that this is still a work-in-progress, it is not yet fully finalized, most features are working but may be unstable, and some options may be missing.
Next steps would be:
- Timings calculations (done) ✅
- Make stop_recording method (done) ✅
- Create export methods & ffmpeg stuff (done) ✅
- Refine cache cleanups (done) ✅
- Examples (mostly done, requires some refining) ✔🧹
- Documentation (mostly done) ✔🧹
- Audio Merging implementation (done, needs to be moved) ✔🧹
- Potential single-track direct recording support? (have an unstable prototype) 🚧

<!-- What is this pull request for? Does it fix any issues? -->

<!-- Link any issues if applicable.
Use keywords from https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests
-->

<!-- Uncomment based on the type of your changes below -->

<!--
## This is a **Code Change**

- [x] I have tested my changes.
- [] I have updated the documentation to reflect the changes.
- [] I have run `task pyright` and fixed the relevant issues.

TrapDrap committed:
Fix decoder

TrapDrap committed:
Add SSRC Cache

TrapDrap committed:
Add support for TTL cache (will be used in the future for more)

TrapDrap committed:
Create voicerec & __init__.py

TrapDrap committed:
Voice rec templates

TrapDrap committed:
Delete voicerec directory

wrong place

TrapDrap committed:
Create __init__.py and voicerec folder

TrapDrap committed:
Add WIP support for voice recording

Not working yet.

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap commented:
The file structure and placements are not final! 
Comments and documentation have not been done.
I preferred working on the voice recorder feature standalone, touching as little of nextcord's main code as possible.
Once everything is working, I will work on splitting them into different files and merging them with nextcord's main code.
As for copyright, I'm not sure what I'm doing so I kinda just added what I think is right.

TrapDrap committed:
Fix decoding channel error

TrapDrap committed:
File updates. Adding beta support for fully functional .wav & .pcm

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
add export option in stop_recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Update to support all formats with ffmpeg

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
fix export not closing properly

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

teaishealthy commented on nextcord/cache.py:
```diff
@@ -0,0 +1,102 @@
+    def __init__(
+        self, maxsize: Optional[int] = None, ttl: Optional[Union[int, float]] = None
+    ) -> None:
+        self.maxsize = maxsize or float("inf")
+        self.ttl = ttl
+        self.timelink = {}
+        self.working = False
```
```suggestion
        self.maxsize: float = float(maxsize) or float("inf")
        self.ttl: float = float(ttl)
        self.timelink = {}
        self.working: bool = False

```

maxsize will be inf even if 0 is passed, not sure if this matters.

TrapDrap replied:
forgot to mention, this is intentional

teaishealthy commented on nextcord/voicerec/core.py:
```diff
@@ -0,0 +1,338 @@
+    def __init__(self, frames: int) -> None:
+        self.frames: int = frames
+
+    def write_to(self, buffer: Union[BufferedWriter, BytesIO]) -> None:
```
buffer should be typed as `io.BufferedIOBase`

TrapDrap replied:
Thank you, will get to that when I work on it again

TrapDrap replied:
closing, reversed change due to incorrect typing (missing methods used in io.BufferedIOBase

teaishealthy commented on nextcord/voicerec/core.py:
```diff
@@ -0,0 +1,338 @@
+        self.frames: int = frames
+
+    def write_to(self, buffer: Union[BufferedWriter, BytesIO]) -> None:
+        write = buffer.write
+        frames = self.frames
```
why?

TrapDrap replied:
Avoiding dot notation when the variable could be used many times, local variables are faster.

TrapDrap replied:
Resolved due to no longer relevant, implement changed

teaishealthy commented on nextcord/voicerec/core.py:
```diff
@@ -0,0 +1,338 @@
+                f"Arg `temp_type` must be of type TempType not `{type(temp_type)}`"
+            )
+
+    def write(self, bytes) -> None:
```
bytes should be `io.BufferedIOBase`

TrapDrap replied:
(And I also shouldn't overwrite the type bytes, will resolve that next update

TrapDrap replied:
closing, reversed change due to incorrect typing (missing methods used in io.BufferedIOBase

TrapDrap commented on nextcord/cache.py:
```diff
@@ -0,0 +1,102 @@
+from time import time as now
+from typing import Optional, Union
+
+__set_item__ = dict.__setitem__
+__get_item__ = dict.__getitem__
+__get__ = dict.get
+__pop__ = dict.pop
```
Yes, although this avoids dot narration which I tested was faster

EmreTech replied:
Does it really have that big of a speed difference? Using the base classes' methods in this manner makes the code less readable.

TrapDrap replied:
Resolving as this results in a 2x+ significance in speedup. Micro-optimization of cache is beneficial as this will be reused in future PRs on cache reworks, these can be called millions of times per second, twice the speed is a major bump well worth slightly less readability in this case.
A detailed comment has been added to describe the reasoning.

TrapDrap committed:
Update nextcord/cache.py

Co-authored-by: teaishealthy <76410798+teaishealthy@users.noreply.github.com>

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap commented on nextcord/cache.py:
```diff
@@ -0,0 +1,102 @@
+    def __init__(self, maxsize: Optional[Union[float, int]] = None) -> None:
+        if not maxsize:
+            maxsize = float("inf")
+        else:
+            self.maxsize = maxsize
```
Yes thank you, although that isn't used (I copied code from my bot lol) so it shouldn't matter. Will commit either way

TrapDrap replied:
This has been reverted. Max size is supposed to be int, the float type was included as support for infinity

TrapDrap committed:
Update nextcord/cache.py

Co-authored-by: teaishealthy <76410798+teaishealthy@users.noreply.github.com>

TrapDrap commented:
I was also wondering if there's a better way of doing my silence writer, right now I'm relying on calling write multiple times in a while loop with bigger to smaller silence frames, I was wondering whether there's a way to create a bytes-like object that can just write the silence frame x times without having to call write x times. It would also not have to cache the 3.5mb or so of silence frames for it to work. The silence is always the same data, so it would be a lot simpler if there's a better way I can write x amount of this data without creating it in memory (multiplication does that, a silence for a minute or two will result in like 100mb of ram spike)

EmreTech commented on nextcord/voicerec/core.py:
```diff
@@ -0,0 +1,338 @@
+"""
+The MIT License (MIT)
+
+:copyright: (c) 2021-present Nextcord Developers
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
```
```suggestion
# SPDX-License-Identifier: MIT
```
Nextcord does not use this module level docstring anymore. Instead, we use a SPDX license identifier comment.

EmreTech commented on nextcord/voicerec/decrypter.py:
```diff
@@ -0,0 +1,64 @@
+"""
+The MIT License (MIT)
+
+:copyright: (c) 2021-present Tag-Epic
+:copyright: (c) 2021-present Nextcord Developers
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
```
```suggestion
# SPDX-License-Identifier: MIT
```

EmreTech commented on nextcord/voicerec/errors.py:
```diff
@@ -0,0 +1,45 @@
+"""
+The MIT License (MIT)
+
+:copyright: (c) 2021-present Nextcord Developers
+
+:license: See license for more info
+"""
```
```suggestion
# SPDX-License-Identifier: MIT
```

EmreTech commented on nextcord/voicerec/exporters.py:
```diff
@@ -0,0 +1,193 @@
+import os
```
```suggestion
# SPDX-License-Identifier: MIT

import os
```

EmreTech commented on nextcord/voicerec/opus.py:
```diff
@@ -0,0 +1,92 @@
+"""
+The MIT License (MIT)
+
+:copyright: (c) 2015-2021 Rapptz
+:copyright: (c) 2021-present Tag-Epic
+:copyright: (c) 2021-present Nextcord Developers
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
```
```suggestion
# SPDX-License-Identifier: MIT
```

EmreTech commented on nextcord/voicerec/shared.py:
```diff
@@ -0,0 +1,47 @@
+"""
+"""
```
```suggestion
# SPDX-License-Identifier: MIT
```

TrapDrap committed:
Update nextcord/voicerec/core.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

TrapDrap committed:
Update nextcord/voicerec/decrypter.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

TrapDrap committed:
Update nextcord/voicerec/errors.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

TrapDrap committed:
Update nextcord/voicerec/exporters.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

TrapDrap committed:
Update nextcord/voicerec/opus.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

TrapDrap committed:
Update nextcord/voicerec/shared.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

teaishealthy commented on nextcord/voicerec/shared.py:
```diff
@@ -0,0 +1,46 @@
+    WAV = 7
+
+
+format_exports = {
+    Formats.MP3: "export_with_ffmpeg",
+    Formats.MP4: "export_with_ffmpeg",
+    Formats.M4A: "export_with_ffmpeg",
+    Formats.MKA: "export_with_ffmpeg",
+    Formats.MKV: "export_with_ffmpeg",
+    Formats.OGG: "export_with_ffmpeg",
+    Formats.PCM: "export_as_PCM",
+    Formats.WAV: "export_as_WAV",
+}
```
why not map directly to functions?

TrapDrap replied:
> why not map directly to functions?

I don't exactly remember rn, I although know that I considered mapping directly to the functions, I must've had a reason.

TrapDrap replied:
changed to address issue

TrapDrap committed:
Fixing cache resolving memory leak & typing issues

TrapDrap committed:
Core update support custom data handler & better starting silence handling

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Update decrypter.py

TrapDrap committed:
Update errors.py (added more error types)

TrapDrap committed:
Revamp starting silence handling when exporting

No longer export with starting silence added, the length of the starting silence will be a property of the AudioFile instead
`AudioFile.starting_silence.milliseconds` used for combining later with pydub (support adding planned)

TrapDrap committed:
Fix opus exporting from wrong place

TrapDrap committed:
Removed format methods from shared

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
update decrypter: handle no nacl installed

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Update exporters: fix typing for linux subprocess

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
fix KeyError when popping non-existent key

TrapDrap committed:
Merge branch 'nextcord:master' into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Refinement update
Changes to be committed:
	renamed:    nextcord/voicerec/__init__.py -> nextcord/voice_recording/__init__.py
	new file:   nextcord/voice_recording/cleanup.py
	renamed:    nextcord/voicerec/core.py -> nextcord/voice_recording/core.py
	renamed:    nextcord/voicerec/decrypter.py -> nextcord/voice_recording/decrypter.py
	renamed:    nextcord/voicerec/errors.py -> nextcord/voice_recording/errors.py
	renamed:    nextcord/voicerec/exporters.py -> nextcord/voice_recording/exporters.py
	renamed:    nextcord/voicerec/opus.py -> nextcord/voice_recording/opus.py
	renamed:    nextcord/voicerec/shared.py -> nextcord/voice_recording/shared.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Fixed typing error, refactors, & wrong value
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/opus.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Adding some basic documentation. (WIP)
	modified:   nextcord/abc.py
	modified:   nextcord/voice_recording/cleanup.py
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/exporters.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Docs (wip)

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Remove extra imports

TrapDrap committed:
attempt fixing circular import

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap committed:
attempt fixing circular import pt 2

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
attempting fixing docs circular import pt 3

TrapDrap committed:
attempt fix circular import pt 4

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
attempt fix docs circular import pt 5

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
again...

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap commented:
no way!! it worked 💀

TrapDrap committed:
add check to fix leakage of old recording to new

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
fix comment

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
adding some more docs

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Rewrote silence writer to be more elegant

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Add periodic silence write & fix leakage detection

Changes to be committed:
	modified:   nextcord/abc.py
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap commented:
I added periodic silence writes which relieves a huge issue with writing silences long term. This defaults to be on, but the delay can be adjusted. Without this feature, if a user hasn't spoken in 3 minutes, it will try to write all 3 minutes of silence the next time they speak, which could result in massive blockages.

TrapDrap committed:
Edit struct size to better suit periodic writes

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap committed:
Fixed docs + add option to toggle leakage detection

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
typo

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Attempt fix all precommit errors

Changes to be committed:
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/decrypter.py
	modified:   nextcord/voice_recording/exporters.py
	modified:   nextcord/voice_recording/opus.py
	modified:   nextcord/voice_recording/shared.py

TrapDrap committed:
fix remaining pre-commit errors

Changes to be committed:
	modified:   nextcord/voice_recording/exporters.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Actually fix it 😓

Changes to be committed:
	modified:   nextcord/voice_recording/exporters.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

alentoghostflame commented on nextcord/cache.py:
```diff
@@ -0,0 +1,149 @@
+    MISSING = 0
+
+
+_MISSING = CacheMissing.MISSING
```
Why not `MISSING` from `nextcord.utils`?

TrapDrap replied:
> Why not `MISSING` from `nextcord.utils`?
Because in this case the cache is expected to be able to store nextcord's MISSING object, without its own MISSING, it could cause issues with ttl & returning content

alentoghostflame commented on nextcord/cache.py:
```diff
@@ -0,0 +1,149 @@
+        if not self.ttl or self.working:
+            return
+
+        self.working = True
```
I'm confused what this is for? Threading support?

TrapDrap replied:
I'm not sure whether it's necessary to be honest, I had it as an extra layer of insurance. Ensures the method doesn't get called twice at once, if that happens it can cause an error during iteration.

alentoghostflame replied:
Due to no async calls being made and ignoring threading (we discourage threading), `__expire__` shouldn't ever be able to run while it's already running.

TrapDrap replied:
Then I could remove this for a tiny speedup, would it be worth it in case the user decides to access the cache from a thread and it completely breaks the ttl structure?

alentoghostflame replied:
If threading support was important, we'd want to use something like `threading.Lock` to properly make sure we don't have multiple `__expire__`'s running.
Also as is, iterating through `BetterTTLCache().items()` and triggering `__expire__` would cause a Changed Size During Iteration error if `__expire__` hit the `del self[item]` stuff. I think you'd need to have everything use the same lock to prevent that, and tbh throwing a "Does not support threading" docstring in the class seems easier than doing all of that.

TrapDrap replied:
Fair, I will add that next commit

alentoghostflame commented on nextcord/cache.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        Update the timelink for a key.
+        Creates the key anew if it does not exist in the timelink.
+        Make sure the item exists in self, could result in memory leak if used improperly!
```
How significantly does a `if key not in self: raise KeyError` slow this down?

TrapDrap replied:
It wouldn't although it's intentionally designed this way, this method is supposed to be a helper method, and the checks depends on whichever method implements it.

TrapDrap replied:
Added a warning to be more implicit

TrapDrap committed:
Address readability issue using `super()`

If further discussion is made in favour of avoiding super,
this change could be reverted.
I still stand against using super as it is an unnecessary function call
when implicity statement could be used. super should be best reserved for
when there are multiple inheritences, and isn't necessary for this use case.

Changes to be committed:
	modified:   nextcord/cache.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

alentoghostflame commented on nextcord/cache.py:
```diff
@@ -0,0 +1,141 @@
+            for item in self:  # iter to delete first item
+                del self[item]
+                break
+        super().__setitem__(key, val)
```
I know you just changed it, but if speed is an issue would doing `dict.__setitem__` adequately speed things up?

TrapDrap replied:
> I know you just changed it, but if speed is an issue would doing `dict.__setitem__` adequately speed things up?

Yes, `dict.__setitem__` is significantly faster than `super()`, only very slightly behind a relay variable due to attribute lookup, it would be a preferred approach if I could.

teaishealthy commented on nextcord/voice_recording/opus.py:
```diff
@@ -0,0 +1,65 @@
+                opus_frame: OpusFrame = self.decode_queue.pop(0)
+
+            except IndexError:
+                sleep(0.001)
```
What and why?

TrapDrap replied:
It is a while loop after all, sleeping when waiting for an item in the queue since the queue isn't expected to receive items more than that often. A frame only arrives every 20ms per user, sleeping during the wait can lower CPU usage.

TrapDrap replied:
Also, the most of opus.py is copied from the previous PR

TrapDrap replied:
Although there is probably a more elegant approach to waiting for an item in queue

teaishealthy replied:
make the decode_queue an actual `queue.Queue` and use it's methods

teaishealthy commented on nextcord/voice_recording/shared.py:
```diff
@@ -0,0 +1,36 @@
+    WAV = 7
+
+
+def open_tmp_file(guild_id, user_id, mode):
+    path = Path(f"{getcwd()}/{TMP_DIR}/{guild_id}.{user_id}.tmp")
+
+    try:
+        return open(path, mode)  # noqa: SIM115
+
+    except FileNotFoundError:
+        makedirs(dirname(path), exist_ok=True)  # Create missing dirs
+        path.touch(exist_ok=True)  # create missing file
+        return open(path, mode)  # noqa: SIM115
```
https://docs.python.org/3/library/tempfile.html

TrapDrap replied:
Thank you this, I will use this, don't have to worry about handling deletion pog

TrapDrap replied:
Okay so I tried to implement this but it will not work as I rely on the names of temp files to access and overwrite them. Using python's built-in tempfile does not allow for naming the file on creation.

teaishealthy commented on nextcord/voice_recording/opus.py:
```diff
@@ -0,0 +1,65 @@
+
+
+class DecoderThread(Thread, nc.opus._OpusStruct):
+    def __init__(self, recorder) -> None:
```
types

teaishealthy commented on nextcord/voice_recording/errors.py:
```diff
@@ -0,0 +1,45 @@
+# SPDX-License-Identifier: MIT
+
+import nextcord.errors as nc_errors
```
why?

TrapDrap replied:
import as prevents circular imports because it waits for the module to be ready before continuing

teaishealthy reviewed (changes requested):
a lot of types are missing, reviewing again when fixed 🦅🦅🦅🦅

TrapDrap committed:
Address no thread support in cache docs

Changes to be committed:
	modified:   nextcord/cache.py

TrapDrap committed:
Specify warning to update_time for TTL cache

Changes to be committed:
	modified:   nextcord/cache.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Resolve missing types in opus.py

Changes to be committed:
	modified:   nextcord/voice_recording/opus.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap commented:
> a lot of types are missing, reviewing again when fixed 🦅🦅🦅🦅

Fixed

TrapDrap committed:
Added better granular control over exporting

Allows for exporting a single item with an AudioWriter instance rather
than only allowing exporting all with AudioData instance

Changes to be committed:
	modified:   nextcord/voice_recording/exporters.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Refactor last commit & Add `export` to AudioWriter

Changes to be committed:
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/exporters.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Fix typing for `export` on AudioWriter

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap committed:
Document AudioWriter & export methods

Changes to be committed:
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/exporters.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Remove deprecated kwarg

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap committed:
Distinct between empty recording & missing tmpfile
Changes to be committed:
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/errors.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Changes to be committed:
	modified:   setup.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Fix decoder not created w/ raw data handler set

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap committed:
Add error: forbid setting handler whilst recording
Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
remove redundant `return None`s

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
fix storing thread (accidentally removed it prior)
Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

TrapDrap committed:
Reverted removing redundant `return None`

cuz apparently its done by style, didnt know wops!

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Add block & allowlist to filter. Filters now work with custom handlers.

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
pre-commit fixes + docs

Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
pre-commit fixes + docs for RecordingFilter
Changes to be committed:
	modified:   nextcord/voice_recording/core.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
exporter docs fixes

Changes to be committed:
	modified:   nextcord/voice_recording/exporters.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
refine error hierarchy

Changes to be committed:
	modified:   nextcord/voice_recording/errors.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Fix not actually setting `OpusFrame.user_id`

TrapDrap committed:
attempting conflict resolving

TrapDrap committed:
Merge branch 'nextcord:master' into TrapDrap-nextcord-voice-recording

TrapDrap committed:
formatting | preparation for examples
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/errors.py
	modified:   nextcord/voice_recording/exporters.py
	modified:   nextcord/voice_recording/opus.py

TrapDrap committed:
multiple fixes + examples (see description)
- add examples & audio merging
    - merging does not work with .wav yet dont know why
    - merging needs to be moved from example into exporters
- add option `prevent_leakage` to `Connectable.connect`
- refine prevent leakage to properly account for latency
- move ffmpeg formats dictionary to shared

	new file:   examples/voice_recording.py
	modified:   nextcord/abc.py
	modified:   nextcord/voice_recording/core.py
	modified:   nextcord/voice_recording/exporters.py
	modified:   nextcord/voice_recording/shared.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
fix precommit errors
Changes to be committed:
	modified:   examples/voice_recording.py

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
fix sending non-merged track as merged
Changes to be committed:
	modified:   examples/voice_recording.py

TrapDrap committed:
change typing for previous fix
Changes to be committed:
	modified:   examples/voice_recording.py

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap committed:
Merge branch 'TrapDrap-nextcord-voice-recording' of https://github.com/TrapDrap/_nextcord into TrapDrap-nextcord-voice-recording

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

teaishealthy reviewed (changes requested):
I think the way you're using (reimplementing) temp files is fundamentally flawed. Instead of this whole `tmp_type` junk and needing to overwrite filenames, keep a lookup table of buffers (!) and let the user bring their own buffer. If it walks like a buffer, it is a buffer. When you actually need to encode with a muxer that needs to be able to seek back, check if you got a file buffer, if not write your data to real temp file and then call ffmpeg.

EmreTech commented on nextcord/voice_recording/opus.py:
```diff
@@ -0,0 +1,65 @@
+    from .core import OpusFrame, RecorderClient
+
+
+class DecoderThread(Thread, nc.opus._OpusStruct):
```
Could we avoid using threads here?

Admiralmatt commented:
Would this feature be able to save to a file as it's recording or does it save in RAM as its recording? if the bot crashes partway through a (6 hour) recording the recording would be lost if it is not saving a temp file to disk.

EmreTech commented on nextcord/cache.py:
```diff
@@ -0,0 +1,151 @@
+# SPDX-License-Identifier: MIT
```
Now conflicts with #1185 because of this.

TrapDrap replied:
I feel both modules should be merged. #1185 is still a draft after all, I don't feel this is a change that should be made in this PR.

TrapDrap committed:
Change Enum comparison from == to is

Co-authored-by: teaishealthy <76410798+teaishealthy@users.noreply.github.com>

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

TrapDrap commented on nextcord/voice_recording/core.py:
```diff
@@ -0,0 +1,1174 @@
+        )
+
+
+# slots only work on python 3.10 with dataclasses
+@dataclass(**{"slots": True} if PYTHON_VERSION >= (3, 10) else {})
+class OpusFrame:
```
Can't hardcode slots, causes an error if we want to support python versions earlier than 3.10

TrapDrap commented on nextcord/voice_recording/core.py:
```diff
@@ -0,0 +1,1174 @@
+        # processes
+        self.process: Optional[Thread] = None
+        self.auto_deaf: bool = auto_deaf if isinstance(auto_deaf, bool) else True
+        self.tmp_type: TmpType = tmp_type or TmpType.File
```
tmp_type can be `None` in which case, default to using `TmpType.File`

TrapDrap committed:
Remove redundant bool call for `self.auto_deaf`

Co-authored-by: teaishealthy <76410798+teaishealthy@users.noreply.github.com>

TrapDrap committed:
call bool instead of `or False`

Co-authored-by: teaishealthy <76410798+teaishealthy@users.noreply.github.com>

TrapDrap commented on nextcord/voice_recording/exporters.py:
```diff
@@ -0,0 +1,371 @@
+    return f
+
+
+class FFmpeg:
+    """A utility class containing audio conversion methods using FFmpeg"""
+
+    @staticmethod
+    def memory_tmp_conv(audio_format: str, writer: AudioWriter) -> bytes:
+        try:
+            return subprocess.Popen(
+                f"{ffmpeg_default_arg} {audio_format} pipe:1",
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                creationflags=FLAG,
+            ).communicate(_read_and_delete(writer.buffer))[0]
+        except FileNotFoundError as e:
+            raise NoFFmpeg(
+                "FFmpeg is not installed or aliased improperly. Unable to launch `ffmpeg` command."
+            ) from e
+
+    @staticmethod
+    def file_tmp_conv(audio_format: str, writer: AudioWriter) -> None:
+        try:
+            data = _read_and_delete(writer.buffer)  # must read first to delete file
+            process = subprocess.Popen(
+                f"{ffmpeg_default_arg} {audio_format} {TMP_DIR}/{writer.guild_id}.{writer.user_id}.tmp",
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                creationflags=FLAG,
+            )
+            process.communicate(data)
+        except FileNotFoundError as e:
+            raise NoFFmpeg(
+                "FFmpeg is not installed or aliased improperly. Unable to launch `ffmpeg` command."
+            ) from e
+
```
That's true, I guess it could be a module. Will move into a module instead.

TrapDrap commented on nextcord/abc.py:
```diff
@@ -1771,15 +1807,19 @@ async def connect(
 
         client = state._get_client()
-        voice = cls(client, self)
+        if recordable:
+            voice = RecorderClient(
+                client, self, auto_deaf, tmp_type, filters, periodic_write_seconds, prevent_leakage
+            )
```
Users should be able to import RecorderClass and superclass it, the implementation in abc.py is just for native recording support like the VoiceClient natively supporting play.

I will look into your suggestion though, see if that's plausible without causing any breaking changes.

TrapDrap commented on nextcord/voice_recording/opus.py:
```diff
@@ -0,0 +1,65 @@
+from time import sleep
+from typing import TYPE_CHECKING, Dict
+
+import nextcord as nc  # `import as` to prevent circular import
```
> I don't get the logic behind this? It makes sense in the errors module, but not here...

To be honest I don't know lol. There was a circular import issue and this solved it.

TrapDrap commented:
> Would this feature be able to save to a file as it's recording or does it save in RAM as its recording? if the bot crashes partway through a (6 hour) recording the recording would be lost if it is not saving a temp file to disk.

This allows you to save to file or ram as it's recording. You can also decide to intercept and process the audio packets yourself in your own manner. If say you recorded using RAM tmp files, the bot crashing will cause the files to be lost. If you record using File tmp files, the bot crashing will keep all the recorded pcm files, but you will need extra functions yourself to process the data into usable recording files.

TrapDrap commented on nextcord/voice_recording/core.py:
```diff
@@ -0,0 +1,1174 @@
+
+    def __init__(
+        self,
+        client=None,
```
(Has been a long time) But if I remember correctly, I wasn't able to typehint this because the class `client` takes (i.e. RecorderClient) was defined after, while the class also needs to call it in its functions so I can't just move it. If you know how I could typehint it, let me know.

TrapDrap commented on nextcord/voice_recording/shared.py:
```diff
@@ -0,0 +1,57 @@
+
+
+ffmpeg_formats = {
+    Formats.MP3: ("mp3", "mp3"),
+    Formats.MP4: ("mp4", "mp4"),
+    Formats.M4A: ("m4a", "ipod"),
+    Formats.MKA: ("mka", "matroska"),
+    Formats.MKV: ("mkv", "matroska"),
+    Formats.OGG: ("ogg", "ogg"),
+    "mp3": ("mp3", "mp3"),
+    "mp4": ("mp4", "mp4"),
+    "m4a": ("m4a", "ipod"),
+    "mka": ("mka", "matroska"),
+    "mkv": ("mkv", "matroska"),
+    "ogg": ("ogg", "ogg"),
```
This is used in the FFMPEG export. It's used to translate the enum into the string used by FFMPEG to dictate the output format.

ooliver1 commented:
As per the description, this is in-progress so will be marked as draft.
